File: Greedy/구명보트.py
from collections import deque
from itertools import combinations


def solution(people, limit):
    # 최소 보트의 개수 = 최소 몸무게 + 최대 몸무게
    answer = 0
    people.sort()
    boat = deque(people)

    # 한명일 경우
    if len(people) == 1:
        return 1

    while boat:
        # 하나 남았을 경우 -> 홀수 일 경우
        if len(boat) == 1:
            answer += 1
            break

        if boat[0] + boat[-1] <= limit:
            answer += 1
            boat.pop()
            boat.popleft()
        else:
            answer += 1
            boat.pop()

    return answer

File: 프로그래머스/2/42885. 구명보트/구명보트.java
import java.util.*;

class Solution {
    public int solution(int[] people, int limit) {
        Arrays.sort(people);
        
        int left = 0;
        int right = people.length - 1;
        int answer = 0;

        while (left <= right) {
            answer++; // 보트 1대는 무조건 사용 (right는 항상 탑승)

            if (people[left] + people[right] <= limit) {
                left++; // 가벼운 사람도 같이 태울 수 있으면 태움
            }
            right--; // 무거운 사람은 항상 처리
        }
        
        return answer;
    }
}
